helper/network-server-helper.cc

```diff
@@ -86,28 +82,34 @@ Ptr<Application>
 NetworkServerHelper::InstallPriv(Ptr<Node> node)
 {
     NS_LOG_FUNCTION(this << node);
+    NS_ASSERT_MSG(node->GetNDevices() > 0, "No gateways connected to provided node");
 
     Ptr<NetworkServer> app = m_factory.Create<NetworkServer>();
 
     app->SetNode(node);
     node->AddApplication(app);
 
-    // Cycle on each gateway
-    for (auto i = m_gateways.Begin(); i != m_gateways.End(); i++)
-    {
-        // Add the connections with the gateway
-        // Create a PointToPoint link between gateway and NS
-        NetDeviceContainer container = p2pHelper.Install(node, *i);
-
-        // Add the gateway to the NS list
-        app->AddGateway(*i, container.Get(0));
-    }
-
-    // Link the NetworkServer to its NetDevices
     for (uint32_t i = 0; i < node->GetNDevices(); i++)
     {
+        // Link the NetworkServer app to its NetDevices
         Ptr<NetDevice> currentNetDevice = node->GetDevice(i);
         currentNetDevice->SetReceiveCallback(MakeCallback(&NetworkServer::Receive, app));
+
+        // Register gateways
+        Ptr<Channel> channel = currentNetDevice->GetChannel();
+        NS_ASSERT_MSG(DynamicCast<PointToPointChannel>(channel),
+                      "Connection with gateways is not PointToPoint");
+        for (uint32_t j = 0; j < channel->GetNDevices(); ++j)
+        {
+            Ptr<Node> gwNode = channel->GetDevice(j)->GetNode();
```

Comment:
If I understand correctly, you are looping through the channel endpoint devices (which should be 2 given that it is a P2P channel?). Therefore, this `gwNode` is not necessarily the gateway, but it could be either the actual gateway, or the network server node. If this is correct, please rename the corresponding variables to reflect this.                                                            Anyway, I am not a huge fan of this solution since 1) it assumes that the link is indeed P2P, which seems to be an assumption which will not hold with future developments and 2) it is quite convoluted. What about adding a `NetworkServerHelper` method that accepts as parameters the gateway nodes and the corresponding P2P devices, and stores these pairs to then pass to `AddGateway` ? For instance, something like` RegisterGatewayP2P(Node gw, NetDevice p2p)` to be called when installing the P2P link

Reply:
Ideally, the server should not have to retrieve the gateway's node. This happens because the packet scheduling features that in real networks happen in gateways, here are implemented in the server. So, currently the server - with a notable abstraction jump - directly gets the gateway's MAC layer from the node to access information (mainly, availability for next transmission) used to schedule a packet. In reality, the server should send packets to gateways with an attached timestamp and gateways should take care of scheduling the radio transmission just-in-time.

My line of though was to try and automate the GW node's retrieval, hiding stuff that may be changed in the future. That said, I am conflicted on how to proceed. On one hand, I like your solution because it makes explicit from a user perspective that we currently only support P2P. On the other, I do not like passing the NetDevice, as it is an object already accessible through the NS node... Mind that we are talking of **the P2P net device of the NS** connected to the GW we are passing the node of, not the one of the GW. It seems to me that with `RegisterGatewayP2P(Node gw, NetDevice p2p)` we would just move the convoluted part in the main.cc file.

For clarity: it looked cleaner beforehand because the P2P connection was instantiated by the NS helper, thus having GW's node and NS's P2P net device in the same place. This was causing the bug described on top of this thread, because users did not know where the P2P instantiation was happening. It gets convoluted because we are trying to retrieve something we are not supposed to (the GW's node, and the MAC later on) from the NS.